akka-http-core/src/main/scala/akka/http/scaladsl/model/MediaType.scala

```diff
@@ -5,445 +5,411 @@
-   * Create a custom media type.
-   */
-  def custom(mainType: String, subType: String, encoding: MediaType.Encoding,
-             compressible: Boolean = false, fileExtensions: immutable.Seq[String] = Nil,
-             params: Map[String, String] = Map.empty, allowArbitrarySubtypes: Boolean = false): MediaType = {
-    require(mainType != "multipart", "Cannot create a MultipartMediaType here, use `multipart.apply` instead!")
-    require(allowArbitrarySubtypes || subType != "*", "Cannot create a MediaRange here, use `MediaRange.custom` instead!")
-    val r = new StringRendering ~~ mainType ~~ '/' ~~ subType
-    if (params.nonEmpty) params foreach { case (k, v) ⇒ r ~~ ';' ~~ ' ' ~~ k ~~ '=' ~~# v }
-    new NonMultipartMediaType(r.get, mainType, subType, compressible, encoding, fileExtensions, params) {
-      override def isApplication = mainType == "application"
-      override def isAudio = mainType == "audio"
-      override def isImage = mainType == "image"
-      override def isMessage = mainType == "message"
-      override def isText = mainType == "text"
-      override def isVideo = mainType == "video"
-    }
+  sealed abstract class NonMultipartWithOpenCharset(val value: String, val mainType: String, val subType: String,
+                                                    val fileExtensions: List[String]) extends WithOpenCharset {
+    def params: Map[String, String] = Map.empty
+    def withParams(params: Map[String, String]): WithOpenCharset with MediaType =
+      customWithOpenCharset(mainType, subType, fileExtensions, params)
   }
 
-  def custom(value: String, encoding: MediaType.Encoding): MediaType = {
-    val parts = value.split('/')
-    if (parts.length != 2) throw new IllegalArgumentException(value + " is not a valid media-type")
-    custom(parts(0), parts(1), encoding)
+  final class Multipart(val subType: String, val params: Map[String, String])
+    extends WithOpenCharset with jm.MediaType.Multipart {
+    val value = renderValue(mainType, subType, params)
+    override def mainType = "multipart"
+    override def isMultipart = true
+    override def fileExtensions = Nil
+    def withParams(params: Map[String, String]): MediaType.Multipart = new MediaType.Multipart(subType, params)
+    def withBoundary(boundary: String): MediaType.Multipart =
+      withParams(if (boundary.isEmpty) params - "boundary" else params.updated("boundary", boundary))
   }
 
-  /**
-   * Tries to parse a ``MediaType`` value from the given String. Returns ``Right(mediaType)`` if successful and
-   * ``Left(errors)`` otherwise.
-   */
-  def parse(value: String): Either[List[ErrorInfo], MediaType] =
-    ContentType.parse(value).right.map(_.mediaType)
+  sealed abstract class Compressibility(val compressible: Boolean)
+  case object Compressible extends Compressibility(compressible = true)
+  case object NotCompressible extends Compressibility(compressible = false)
+  case object Gzipped extends Compressibility(compressible = false)
```

Comment:
Right, I see.

I guess when rolling my own MediaType (and it's compressed using something) I could use `NotCompressible`, but somehow feels that this could be useful to allow users to extend (i.e. drop the `sealed` off `Compressability`) - let's say I want to add my special content-encoding, in a similar way like we do with the built-in gzip.